xds/src/test/java/io/grpc/xds/internal/sds/SdsProtocolNegotiatorsTest.java

```diff
@@ -259,10 +271,9 @@ public void clientSdsProtocolNegotiatorNewHandler_fireProtocolNegotiationEvent()
   }
 
   @Test
-  public void serverSdsProtocolNegotiator_passNulls_expectPlaintext() {
+  public void serverSdsProtocolNegotiator_nullSyncContext_expectPlaintext() {
     InternalProtocolNegotiator.ProtocolNegotiator protocolNegotiator =
-        SdsProtocolNegotiators.serverProtocolNegotiator(null, 7000,
-            null);
+        SdsProtocolNegotiators.serverProtocolNegotiator(7000, null);
```

Comment:
can you use? `/* paramName= */`, like i mentioned before this is actually checked by linter which is less error prone. applies to other places.

Reply:
I was talking to @ejona86 some time ago and he said for test code this is completely unnecessary (IIRC). It will be good to have consensus

Reply:
I don't think we have an exception for this for test code, but I also don't see any failure in the build so I don't think lint complains about this.

Looking at ErrorProne these are the checks I could find that are impacted by argument comments (none of them apply):
https://errorprone.info/bugpattern/ArgumentSelectionDefectChecker
https://errorprone.info/bugpattern/BooleanParameter
https://errorprone.info/bugpattern/ParameterComment
https://errorprone.info/bugpattern/ParameterName

It isn't at all clear to me why `syncContext` would be `null`, and why that would imply plain-text. That seems super-strange to me. The `port` also seems unnecessary since the negotiator will have access to that information from io.netty.channel.Channel.localAddress().

I don't think we have a rule that you are required to put the parameter names, but you should put the parameter names when it could be confusing otherwise. In this case the `null` is really important to the test, but there's also only one null. The previous version is more confusing. In any case for a test, instead of having the comment I would be more prone to doing:
```
SynchronizationContext syncContext = null; // or syncContext = null; if it is a class field
SdsProtocolNegotiators.serverProtocolNegotiator(7000, syncContext);
```

But really, I'd hope we can drop the 7000 and always specify a syncContext.

Reply:
Hmm... I guess the 7000 is specified here since there will be a single watch for the entire server. This is a problem though, because that is not normal initialization for a protocol negotiator; this approach won't be available in a ServerCredentials API. There may need to be more hole-punching to allow it.

Reply:
Yes the port is provided for setting the LDS watch for the server (not sure what you mean by "entire server"). ServerCredentials is attached to a server and a server is attached to a creds (since you can't reuse creds?) In that case you can get the server port from the creds, right?

Reply:
`syncContext` can never be null since it is created by the server builder before invoking the SdsProtocolNegotiator.  So the test is a bit misleading but is there to simulate XdsClientImpl throwing exception (because syncContext is null).

Reply:
added the param comments after doing a quick search

Reply:
The user will create any ServerCredentials before they are passed to the ServerBuilder, so [passing in port like this simply won't work](https://github.com/grpc/grpc-java/blob/54f9e29cdb205062158e19d9dc1a567afd61a318/xds/src/main/java/io/grpc/xds/internal/sds/XdsServerBuilder.java#L177). And yes, the user could use the same server credentials with multiple servers. It also looks like you are having "creation imply starting", as the credential appears to start the watch when it is created. The user will create the credential well before the port is even determined, so that is also not going to work. These things are important to call out because they mean additional plumbing would be required.

This test is horrible. Abusing `syncContext` in that way would deserve a big comment or _some_ sort of pointer to something weird going on. Naming it something like "serverSdsProtocolNegotiator_xdsClientCreationThrows_expectPlaintext" would be a good starter. But that is even wrong. `syncContext` is a red herring because [the only catch in SdsProtocolNegotiators is for IOException](https://github.com/grpc/grpc-java/blob/54f9e29cdb205062158e19d9dc1a567afd61a318/xds/src/main/java/io/grpc/xds/internal/sds/SdsProtocolNegotiators.java#L289). That means it wouldn't catch NullPointerException and it also wouldn't catch [NoSuchElementException](https://github.com/grpc/grpc-java/blob/54f9e29cdb205062158e19d9dc1a567afd61a318/xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java#L86). So at the very least I've determined the exception handling is broken (XdsClientWrapperForServerSds should probably not be throwing a RuntimeException based on bad input from a user). The reason the test passes is because it triggers "java.io.IOException: Environment variable GRPC_XDS_BOOTSTRAP not defined.".

What is clearly missing is calling `serverProtocolNegotiator()` and it returning an actual ServerSdsProtocolNegotiator. That said, even ServerSdsProtocolNegotiator should have a protocol of `http` as the xds server may not return security information. (Actually, it seems it should delegate to a fallback protocol negotiator, e.g., serverPlaintext(), so the protocol wouldn't end up changing.)

Reply:
> The user will create the credential well before the port is even determined, so that is also not going to work. These things are important to call out because they mean additional plumbing would be required.

But the port is determined or known when the server is built (`XdsServerBuilder.build`) and that is when the watch is set: although I agree it could be delayed until `start`. Doing that will require me to reintroduce `XdsServer` I had in a previous PR for the shutdown case.  Can be done in a separate PR.

> But that is even wrong. syncContext is a red herring because the only catch in SdsProtocolNegotiators is for IOException. That means it wouldn't catch NullPointerException and it also wouldn't catch NoSuchElementException. So at the very least I've determined the exception handling is broken (XdsClientWrapperForServerSds should probably not be throwing a RuntimeException based on bad input from a user).

Interesting - I did have the broader exception earlier but made it narrower based on PR comments. Again I can fix this in a separate PR after we agree on the approach.

> What is clearly missing is calling serverProtocolNegotiator() and it returning an actual ServerSdsProtocolNegotiator. 

So you think returning `InternalProtocolNegotiators.serverPlaintext()` is not correct for `serverProtocolNegotiator()` and should always be delegated to the `ServerSdsProtocolNegotiator`? I think that was the old approach.

> That said, even ServerSdsProtocolNegotiator should have a protocol of http as the xds server may not return security information.

Interesting. The `ServerSdsProtocolNegotiator` does the right thing based on the current state of the security information from the xDS server (whether DownstreamTlsContext is present or not) just that SCHEME is hardcoded to https. If you are saying it should always be "http" that should be easy to change. Changing it to "https" or "http" based on the current state is also not that hard (I think).

> (Actually, it seems it should delegate to a fallback protocol negotiator, e.g., serverPlaintext(), so the protocol wouldn't end up changing.)

It does delegate to `serverPlaintext()` at the time of creating the handler if you see `HandlerPickerHandler.newHandler`.

Reply:
> But the port is determined or known when the server is built (XdsServerBuilder.build) and that is when the watch is set: although I agree it could be delayed until start.

Even now it must be delayed until start, because that's when I/O is permitted to start. Technically, the user could create a server and never start it. Yes, I'd expect fixing this sort of thing to be done in a separate PR.

> Interesting - I did have the broader exception earlier but made it narrower based on PR comments. Again I can fix this in a separate PR after we agree on the approach.

The problem was with the code throwing a RuntimeException. The reviewer was right to ask for more narrow exception handling. But code randomly throwing RuntimeExceptions is asking for bugs, like failure to notice changing Exception to IOException in a catch would leave some cases unhandled. In our code, RuntimeExceptions should generally be reserved for programmer errors.

> So you think returning InternalProtocolNegotiators.serverPlaintext() is not correct for serverProtocolNegotiator() and should always be delegated to the ServerSdsProtocolNegotiator? I think that was the old approach.

If the bootstrap is unavailable, returning serverPlaintext is fine. But that's "not the end of it." I thought we had decided we were going to have the user specify a fallback credential in case xds has no tls configuration, so it seems some credential should still be provided to ServerSdsProtocolNegotiator as a fallback and that should behave the same from the user's perspective independent of when it happens. (i.e., the user should see the same behavior if GRPC_XDS_BOOTSTRAP is unset as if the xds server has no tls configuration)

> Interesting. The ServerSdsProtocolNegotiator does the right thing based on the current state of the security information from the xDS server (whether DownstreamTlsContext is present or not) just that SCHEME is hardcoded to https. If you are saying it should always be "http" that should be easy to change. Changing it to "https" or "http" based on the current state is also not that hard (I think).

Does it guarantee it is going to use a secure connection? If not, it should be http. So yes, it should be http.

Changing it on-the-fly won't work, because this is called separately from an actual connection being handled and the setting can change at any point based on xds. There's no form of synchronization available to make sure that the right value is used for the right connections.

(In truth, this value is unused on server-side today. So it doesn't really matter. Except this is security stuff. So it totally matters.)

> It does delegate to serverPlaintext() at the time of creating the handler if you see HandlerPickerHandler.newHandler.

It should have a "fallback" protocol negotiator passed in. That will mirror the final design where the user gets to decide. Hard-coding this stuff now means that we're less likely to notice issues with the long-term design.

Reply:
@ejona86 for all of your comments requiring code change I'll create a separate PR instead of complicating this PR (which was meant to remove DownstreamTlsContext from the XdsServerBuilder). For "fallback" too the XdsServerBuilder could define something like say `XdsServerBuilder.fallbackProtocolNegotiator(ProtocolNegotiator protocolNegotiator)` and plumb it down. Again let's do it in a separate PR or at the time of XdsServerCredentials.

I can create an issue to keep track of all items needing that separate PR.

Reply:
And there is one case of `NoSuchElementException` that I can convert to IOException or log message and return `null` (in another PR).